### PR TITLE
Optimize MergeBlocks by caching branch results

### PR DESCRIPTION
**Pull request by kripken**
`BranchSeekerCache` caches the set of branches in a node +
its children, and helps compute new results by looking in the cache
and using data for the children. This avoids quadratic time in the
common case of a post-walk on a tower of nested blocks which is
common in a switch.

Fixes https://github.com/WebAssembly/binaryen/issues/3090 . On the testcase there this pass goes from
over a minute to less than a second.

I haven't looked at the validator yet which may also be helped by this,
but as this ended up more work than I thought I didn't want to do too
much in one PR.

Fuzzed for a few thousand iterations + random50 on emscripten test suite.

### CODE REVIEW

File: src/passes/MergeBlocks.cpp [hunk 1]
```diff
@@ -360,6 +365,13 @@ optimizeBlock(Block* curr, Module* module, PassOptions& passOptions) {
         if (loop) {
           loop->finalize();
         }
+        // Note that we modify the child block here, which invalidates info
+        // in branchInfo. However, as we have scanned the parent, we have
+        // already forgotten the child's info, so there is nothing to do here
+        // for the child.
+        // (We also don't need to do anything for the parent - we move code
+        // from a child into the parent, but that doens't change the total
```
Comment (tlively) [inline src/passes/MergeBlocks.cpp:373]:
```suggestion
        // from a child into the parent, but that doesn't change the total
```

File: src/ir/branch-utils.h [hunk 2]
```diff
@@ -218,6 +233,77 @@ struct BranchSeeker : public PostWalker<BranchSeeker> {
...
+      // This helps the case of the first child being a block with a very large
+      // set of names.
+      if (currBranches.empty()) {
+        currBranches.swap(moreBranches);
+      } else {
+        currBranches.insert(moreBranches.begin(), moreBranches.end());
+      }
+    };
+    // Add from the children, which are hopefully cached.
+    for (auto child : ChildIterator(curr)) {
```
Comment (tlively) [inline src/ir/branch-utils.h:276]: Why do we need to look at children but not more remote descendants?

Comment (kripken) [reply src/ir/branch-utils.h:276]: The data for each child in `branches[child]` contains all the grandchildren etc. as well.

Likewise if the data isn't cached, `BranchAccumulator` below will scan into grandchildren etc.

Comment (tlively) [reply src/ir/branch-utils.h:276]: aha, I get it now 👍

**Review by tlively — APPROVED** [review]

**Pull request merged**